internal/serverinstall/k8s.go

```diff
@@ -263,7 +267,259 @@ func (i *K8sInstaller) Install(
+
+		namespaceClient := clientset.CoreV1().Namespaces()
+		_, err := namespaceClient.Get(context.TODO(), "openshift", metav1.GetOptions{})
+		isOpenShift := err == nil
+
+		// Default namespace in OpenShift acts like a regular K8s namespace, so we don't want
+		// to remove fsGroup in this case.
+		if isOpenShift && i.config.namespace != "default" {
+			s.Update("OpenShift detected. Switching configuration...")
+			i.config.openshift = true
+		}
+	}
+
+	s.Done()
+
+	statefulSetClient := clientset.AppsV1().StatefulSets(i.config.namespace)
+	waypointStatefulSet, err := statefulSetClient.Get(ctx, serverName, metav1.GetOptions{})
+	if err != nil {
+		ui.Output(
+			"Error obtaining waypoint statefulset: %s", clierrors.Humanize(err),
+			terminal.WithErrorStyle(),
+		)
+		return nil, err
+	}
+
+	s = sg.Add("Upgrading server to %q", i.config.serverImage)
+
+	// Update pod image to requested serverImage
+	cmd := exec.Command("kubectl", "set", "image", "statefulset", serverName, fmt.Sprintf("server=%s", i.config.serverImage))
+	cmd.Stdout = s.TermOutput()
+	cmd.Stderr = cmd.Stdout
+
+	if err = cmd.Run(); err != nil {
+		ui.Output(
+			"Error executing kubectl to update server image in pod: %s", clierrors.Humanize(err),
+			terminal.WithErrorStyle(),
+		)
+
+		return nil, err
+	}
+
+	if waypointStatefulSet.Spec.UpdateStrategy.Type == "OnDelete" {
+		s.Update("Deleting pod to refresh image")
+		log.Info("Update Strategy is 'OnDelete', deleting pod to refresh image")
+
+		podClient := clientset.CoreV1().Pods(i.config.namespace)
+		if podList, err := podClient.List(ctx, metav1.ListOptions{LabelSelector: fmt.Sprintf("app=%s", serverName)}); err != nil {
+			ui.Output(
+				"Error listing pods: %s", clierrors.Humanize(err),
+				terminal.WithErrorStyle(),
+			)
+			return nil, err
+		} else {
+			for _, pod := range podList.Items {
+				if err := podClient.Delete(ctx, pod.Name, metav1.DeleteOptions{}); err != nil {
+					s.Update("Pod deletion failed", terminal.WithErrorStyle)
+					s.Done()
+					ui.Output(
+						"Error deleting pod %q: %s", pod.Name, clierrors.Humanize(err),
+						terminal.WithErrorStyle(),
+					)
+					return nil, err
+				}
+			}
+		}
+
+		log.Info("Pod(s) deleted, k8s will now restart waypoint server ", serverName)
+	} else if waypointStatefulSet.Spec.UpdateStrategy.Type == "RollingUpdate" {
+		log.Info("Update Strategy is 'RollingUpdate', no further action required")
+	} else {
+		log.Warn("Update Strategy is not recognized, so no action is taken", "UpdateStrategy",
+			waypointStatefulSet.Spec.UpdateStrategy.Type)
+	}
+
+	s.Update("Image set to update!")
+	s.Done()
+
+	s = sg.Add("Waiting for server to be ready...")
+	log.Info("waiting for waypoint server to become ready after image refresh")
+
+	var contextConfig clicontext.Config
+	var advertiseAddr pb.ServerConfig_AdvertiseAddr
+	var httpAddr string
+	var grpcAddr string
+
+	err = wait.PollImmediate(2*time.Second, 1*time.Minute, func() (bool, error) {
```

Comment:
i found the services to be slower than statefulsets at upgrading etc, so i'd bump this to 2 minutes just to be safe
```suggestion
	err = wait.PollImmediate(2*time.Second, 2*time.Minute, func() (bool, error) {
```